shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
-        if (permissionsMap[owner]) {
-          newPermissions.remove(permissionsMap[owner]);
-        }
+    if (vertexId.slice(0, 2) === "o:") {
+      // Owner. We don't need to do anything.
+      incomingEdges = [];
+    } else if (vertexId.slice(0, 2) === "t:") {
+      const token = context.tokensById[vertexId.slice(2)];
+      if (token) {
+        incomingEdges = [tokenToEdge(token)];
+      }
+    } else if (vertexId.slice(0, 2) === "i:") {
+      if (ownerIdentityIds.indexOf(vertexId.slice(2)) >= 0) {
+        const p = new MembranedPermissionSet(PermissionSet.fromRoleAssignment({ allAccess: null },
+                                                                              viewInfo),
+                                             new Clause());
+        incomingEdges = [{ sharerId: "o:Owner", membranedPermissions: p }];
+      } else if (!grain.private) { // legacy public grain
+        const p = new MembranedPermissionSet(PermissionSet.fromRoleAssignment({ none: null },
+                                                                              viewInfo),
+                                             new Clause());
+        incomingEdges = [{ sharerId: "o:Owner", membranedPermissions: p }];
+      } else {
+        incomingEdges = (context.tokensByRecipient[vertexId.slice(2)] || []).map(tokenToEdge);
+      }
+    } else {
+      throw new Meteor.Error(500, "Unrecognized vertex ID: " + vertexId);
+    }
 
-        if (permissionsMap[sharer].add(newPermissions)) {
-          needToPush = true;
+    incomingEdges.forEach((edge) => {
+      const sharerId = edge.sharerId;
+      let needToPush = false;
+      // For each MembranedPermissionSet in permissionsMap[vertexId],
+      // apply edge.membranedPermissions in sequence.
+      const sequenced = {};
+      for (const clauseHash in permissionsMap[vertexId]) {
+        const newPermissions = edge.membranedPermissions.clone();
+        newPermissions.sequence(permissionsMap[vertexId][clauseHash]);
+        const newHash = newPermissions.membrane.hash();
+        if (sequenced[newHash]) {
+          sequenced[newHash].permissions.add(newPermissions.permissions);
+          for (tokenId in newPermissions.tokensUsed) {
+            sequenced[newHash].tokensUsed[tokenId] = true;
+          }
+        } else {
+          sequenced[newHash] = newPermissions;
         }
-        if (needToPush) {
-          userStack.push(sharer);
+      };
+
+      // Optimization: we don't care about permissions that we've already proven the opener has.
+      if (permissionsMap["o:Owner"]) {
+        for (const hashedClause in permissionsMap["o:Owner"]) {
+          const membranedPermissionSet = permissionsMap["o:Owner"][hashedClause];
+          if (hashedClause in sequenced) {
+            const smps = sequenced[hashedClause];
+            smps.permissions.remove(membranedPermissionSet.permissions);
+            if (smps.permissions.array.length == 0) {
```

Comment:
Shouldn't you be checking that all elements in the array are false, rather than that the array is empty? (Maybe add an `isEmpty()` method to `PermissionSet`?)